src/main/java/bsh/Interpreter.java

```diff
@@ -1158,6 +1167,27 @@ public void setErr( PrintStream err ) {
         this.err = err;
     }
 
+    /**
+     * Interrupt the current parsing even if blocked reading from the input
+     * stream. It does not stop a long operation that is being executed.
+     *
+     * Once interrupted the input stream and the parser become invalid and can
+     * not be used any more.
+     *
+     * Override this basic implementation for more sophisticated policy.
+     */
+    protected void interrupt() {
+        println("... interrupted ...");
+
+        try {
+            interrupted = true;
```

Comment:
If we had a factory method here, we could close the is and recreate the parser. Currently we just close the stream which means the parser won't work any more. Subclasses will have to override this method to recover the parser.

Reply:
Can we not use eof instead? In other words, if it is interrupted it is eof.

Reply:
I was using eof at the beginning, but it was the cause why "class A {\n ^C" printed a parser error. I realized that when you close the stream after the parser parsed "why class A {\n", Line() returns still false therefore you go on trying to parse the node and the parser throws a ParserException consuming <EOF> as a token (this is why you needed to parse the error message). Maybe it can be fixed in the grammar... I do not know.